Mathlib/Util/CompileInductive.lean

```diff
@@ -0,0 +1,204 @@
+    type := ← mkEq rv.type old new
+    value := ← forallTelescope rv.type λ xs body => do
+      let pf := .const rv.name (.zero :: levels.tail!)
+      let pf := mkAppN pf xs[:rv.numParams]
+      let old := mkAppN old xs
+      let new := mkAppN new xs
+      let pf := .app pf <| ← mkLambdaFVars xs[rv.getFirstIndexIdx:] <| ← mkEq body old new
+      let minor := xs[rv.getFirstMinorIdx]!
+      let pf := .app pf <| ← forallTelescope (← inferType minor) λ ys body' => do
+        let pf' ← mkEqRefl body' <| mkAppN minor ys
+        mkLambdaFVars ys pf'
+      let pf := .app pf xs[rv.getMajorIdx]!
+      mkFunExts' xs pf (body, old, new)
+  }
+  Compiler.CSimp.add name .global
+  compileDefn <| ← getConstInfoDefn <| mkRecOnName iv.name
+
+/--
+Generate compiled code for the recursor for `iv`.
+-/
+def compileInductive (iv : InductiveVal) : TermElabM Unit := do
+  let rv ← getConstInfoRec <| mkRecName iv.name
+  if ← isProp rv.type then
+    logWarning m!"not compiling {rv.name}"
+    return
+  unless rv.numMotives == 1 do
+    throwError "mutual/nested inductives unsupported"
+  if iv.type.getForallBody.isProp && !iv.isRec && iv.numCtors == 1 && iv.numIndices == 0 then
+    compilePropStruct iv rv
+    return
+  let levels := rv.levelParams.map .param
+  let name ← mkFreshUserName rv.name
+  addAndCompile <| .mutualDefnDecl [{ rv with
+    name
+    value := ← forallTelescope rv.type λ xs body => do
+      let val := .const (mkCasesOnName iv.name) levels
+      let val := mkAppN val xs[:rv.numParams]
+      let val := .app val <| ← mkLambdaFVars xs[rv.getFirstIndexIdx:] body
+      let val := mkAppN val xs[rv.getFirstIndexIdx:]
+      let val := mkAppN val <| rv.rules.toArray.map λ rule =>
+        .beta (replaceConst rv.name name rule.rhs) xs[:rv.getFirstIndexIdx]
+      mkLambdaFVars xs val
+    hints := .opaque
+    safety := .partial
+  }]
+  let old := .const rv.name levels
+  let new := .const name levels
+  let name ← mkFreshUserName <| rv.name.str "eq"
+  addDecl <| .mutualDefnDecl [{
+    name
+    levelParams := rv.levelParams
+    type := ← mkEq rv.type old new
+    value := ← forallTelescope rv.type λ xs body => do
+      let pf := .const rv.name (.zero :: levels.tail!)
+      let pf := mkAppN pf xs[:rv.numParams]
+      let old := mkAppN old xs
+      let new := mkAppN new xs
+      let motive ← mkLambdaFVars xs[rv.getFirstIndexIdx:] <| ← mkEq body old new
+      let pf := .app pf motive
+      let pf := mkAppN pf <| ← rv.rules.toArray.zip xs[rv.getFirstMinorIdx:]
+        |>.mapM λ (rule, minor) => do
+        forallTelescope ((← inferType minor).replaceFVar xs[rv.numParams]! motive) λ ys _ => do
+          let minor := mkAppN minor ys[:rule.nfields]
+          let pf' ← mkEqRefl (← inferType minor) minor
+          let pf' ← ys[rule.nfields:].foldlM (λ pf' y => do mkCongr pf' (← mkFunExts y)) pf'
+          mkLambdaFVars ys pf'
+      let pf := mkAppN pf xs[rv.getFirstIndexIdx:]
+      mkFunExts' xs pf (body, old, new)
```

Comment:
```suggestion
    value := .const name (rv.levelParams.map .param)
```
Not sure how seriously to take this as a suggestion, but it works... Since this declaration is declared as `safety := .partial` it's already working in an inconsistent context and we don't have to prove anything.

Reply:
Oh! I thought that the kernel checks the inhabitedness of the type.  Because the usual story is "partial doesn't compromise soundness".

In that case we might even do away with the csimp charade entirely and just declare it as implemented_by.

Reply:
Yeah, `DefinitionSafety.partial` is really confusing, because it's not the same thing as the `partial` modifier users can put on definitions. `DefinitionSafety.partial` is basically a synonym of "unsafe" as far as the kernel is concerned: you are allowed to do general recursion, but you cannot directly reference it in safe declarations. When you write `partial def foo : A := foo`, lean does `addAndCompile { name := foo', safety := .partial, value := foo' }` to declare the computational version and `addDecl { name := foo, safety := .safe, value := default }` for the model definition, and then connects the two with `implemented_by`.

Reply:
Apparently this is just inconsistent terminology.  Kernel-level partial defs are different than user-level partial defs.  The kernel ones are essentially the same as unsafe, while the user-level ones are perfectly safe and usable in proofs.

```lean
import Lean
open Lean

#eval addDecl <| .mutualDefnDecl [{
  name := `False.intro
  levelParams := []
  type := .const ``False []
  value := .const `False.intro []
  hints := .opaque
  safety := .partial
}]

#check False.intro -- False
#print axioms False.intro -- 'False.intro' does not depend on any axioms

theorem please_dont_typecheck : False := False.intro
-- (kernel) invalid declaration, safe declaration must not contain partial declaration 'False.intro'
```

Reply:
FWIW, `implemented_by` doesn't work across modules.  So let's go with your unsafe proof version.

Reply:
> ```lean
> #print axioms False.intro -- 'False.intro' does not depend on any axioms
> ```

We should really mention when a constant is `partial` or `unsafe` in `#print axioms`...

Reply:
Another variation would be to use `lcProof` and make the equality `.unsafe`. But I guess this version results in a smaller proof term.